### PR TITLE
Better logging

### LINKED ISSUE #6: Incorporate Better Logging into the Application
Currently, the application doesn't have much logging. We should add log output for errors, warnings, etc.

In cases of critical errors, we can even automatically send emails to the admins.

### PR DESCRIPTION
**Pull request by SamLab17**
Adds a logging implementation to the application along with some log messages added in throughout the code.

Log messages with a severity of `FATAL` or those with a severity of `ERROR` plus an optional `notify` argument will be sent to admin emails (as defined in the config file).

A log level can also be specified to mask out some less severe log messages.

## How to Use

To use this logger, make sure it is imported at the top of the file:
```
from helphours import log
```

Then you can write log messages as follows:
```
log.debug('This is a debug message')
log.info('This is an info message')
log.warning('This is a warning message')
log.error('This is an error message)
log.error('This is an error message which will be sent to admins', notify=True)
log.fatal('This is a fatal error message which will also be sent to admins')
```

## Example Usages

https://github.com/SamLab17/HelpHours/blob/65711bfe3a0e177cec88aa73ea9c335acf7e2a10/helphours/error_routes.py#L5-L8

https://github.com/SamLab17/HelpHours/blob/65711bfe3a0e177cec88aa73ea9c335acf7e2a10/helphours/routes.py#L90-L91

https://github.com/SamLab17/HelpHours/blob/f32b0fd0585e5ecb959b6efc4cc0007464e962d9/helphours/account_routes.py#L25

### CODE REVIEW

File: helphours/password_reset.py [hunk 1]
```diff
@@ -20,6 +20,7 @@ def create_reset_request(user):
                           'html')
     expire_time = dt.datetime.utcnow() + dt.timedelta(hours=NUM_HOURS_EXPIRE)
     reset_requests[token] = (expire_time, user.id)
+    log.info(f'{user.first_name} {user.last_name} requested to rest their password.')
```
Comment (ninaahmed) [inline helphours/password_reset.py]: nit, "reset"

**Review by ninaahmed — APPROVED** [review]
Looks great!

File: helphours/logger.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    # The various logging levels to mask out log entries.
+    # e.g. Using LEVEL_WARNING will mask out all DEBUG and INFO messages.
+    LEVEL_DEBUG = 10
+    LEVEL_INFO = 20
+    LEVEL_WARNING = 30
+    LEVEL_ERROR = 40
+    LEVEL_FATAL = 50
+
+    DEBUG_PREFIX = "[Debug]"
+    INFO_PREFIX = "[ Info]"
```
Comment (ClosingBracket) [inline helphours/logger.py:15]: Extra space?

File: helphours/logger.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    WARNING_PREFIX = "[ WARN]"
```
Comment (ClosingBracket) [inline helphours/logger.py:16]: Extra space?

File: helphours/account_routes.py [hunk 4]
```diff
@@ -19,8 +19,10 @@ def login():
                 message = "Incorrect email or password"
             elif not user.is_active:
                 message = "This account is inactive"
+                log.info(f'{user.first_name} attempted to log in, but account was deactivated.')
```
Comment (ClosingBracket) [inline helphours/account_routes.py]: Include last name also to bring into consistency with else, which logs first and last name?

File: helphours/logger.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        self.email = email_notifier
+        self.admin_emails = admin_emails
+        self.log_file = log_file
+        self.log_level = log_level
+
+    # Helper Methods
+    def format_message(severity_prefix, message):
+        now = datetime.now()
+        time_string = now.strftime("[%b %d, %Y][%H:%M:%S]")
+        return f"{time_string}{severity_prefix}: {message}\n"
```
Comment (ClosingBracket) [inline helphours/logger.py:38]: Will there be a space between the time_string and the [INFO], [ERROR] such? If not, we should probably add one.

**Review by ClosingBracket — CHANGES_REQUESTED** [review]

File: helphours/logger.py (hunk 2 above)
Comment (SamLab17) [reply helphours/logger.py:15]: Those were intentional, it's so they all take up the same width. For example:
```
[Aug 16, 2020][15:21:50][Debug]: Example message
[Aug 16, 2020][15:20:12][ Info]: Example message
[Aug 16, 2020][15:21:50][ WARN]: Example message
[Aug 16, 2020][15:21:50][ERROR]: Example message
[Aug 16, 2020][15:21:50][FATAL]: Example message
```
I feel like it makes it look nicer in a file.

File: helphours/logger.py (hunk 3 above)
Comment (SamLab17) [reply helphours/logger.py:16]: (See above)

File: helphours/logger.py (hunk 5 above)
Comment (SamLab17) [reply helphours/logger.py:38]: Up to you if you think we need one. I think it looks alright with the square brackets and no space:
```
[Aug 16, 2020][15:21:50][ERROR]: Example message
```

File: helphours/logger.py (hunk 2 above)
Comment (SamLab17) [reply helphours/logger.py:15]: But we could change it so the space is after the text, e.g. `[Info ]` if that looks better

File: helphours/password_reset.py (hunk 1 above)
Comment (SamLab17) [reply helphours/password_reset.py]: Good catch!

File: helphours/logger.py (hunk 2 above)
Comment (ClosingBracket) [reply helphours/logger.py:15]: Nah, looks good!

**Review by ClosingBracket — APPROVED** [review]
Nice

**Comment by SamLab17** [comment]
I'll merge this PR now since it's pretty mundane but could be very important to have when working on new features.

@Andrew-William-Smith if you see anything that could be changed in this PR, just open an issue for it!

**Pull request merged**